{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Cross-validated classification evaluation\n",
    "\n",
    "The following notebooks describe the evaluation of taxonomy classifiers using cross-validated read data sets. This analysis represents a conventional cross-validated classification, wherein unique sequences are randomly sampled from a reference dataset and used as a test set for taxonomic classification, using a training set that has those sequences removed, but not other sequences that share taxonomic affiliation. Instead, the training set must contain identical taxonomies to those represented by the test sequences.\n",
    "\n",
    "\n",
    "## Structuring new results for comparison to precomputed results\n",
    "To prepare results from another classifier for analysis, you'll need to have tab-delimited taxonomy files that map query sequence IDs to their taxonomy assignments in the format:\n",
    "seqID\ttaxonomy;label;string\n",
    "\n",
    "An example of how to generate these is presented in the [taxonomy assignment notebook](./taxonomy-assignment.ipynb) in this directory, which was used to generated the precomputed data in the [tax-credit repository](https://github.com/caporaso-lab/tax-credit/).\n",
    "\n",
    "Your taxonomy files should be called ``query_tax_assignments.txt``, and nested in the following directory structure:\n",
    "\n",
    "```\n",
    "results_dir/\n",
    " cross-validated/\n",
    "  dataset-id/ \n",
    "    method-id/\n",
    "     parameter-combination-id/\n",
    "      query_tax_assignments.txt\n",
    "```\n",
    "\n",
    "``results_dir`` is the name of the top level directory, and you will set this value in the first code cell of the analysis notebooks. You can name this directory whatever you want to. ``cross-validated`` describes the specific analysis that is being run, and must be named ``cross-validated`` for the framework to find your results.\n",
    "\n",
    "This directory structure is identical to that for the [precomputed results](https://github.com/caporaso-lab/tax-credit/tree/master/data/precomputed-results). You can review that directory structure for an example of how this should look.\n",
    "\n",
    "## Contents\n",
    "\n",
    "The steps involved in preparing and executing novel-taxa analysis are described in a series of notebooks:\n",
    "\n",
    "1) **[Dataset generation](../novel-taxa/dataset-generation.ipynb)** only needs to be performed once for a given reference database. Only run this notebook if you wish to make datasets from a different reference database, or alter the parameters used to make the novel taxa datasets. The default included in Tax-Credit is Greengenes 13\\_8 release, amplified *in silico* with primers 515f and 806r, and trimmed to 250 nt from the 5' end. This notebook is actually included in the ``novel-taxa`` notebook directory, as simulated reads and \"novel taxa\" are generated simultaneously from the same source datasets.\n",
    "\n",
    "2) **[Taxonomic classification](./taxonomy-assignment.ipynb)** of simulated reads is performed using the datasets generated in *step 1*. This template currently describes classification using QIIME 1 classifiers and can be used as a template for classifiers that are called via command line interface. Python-based classifiers can be used following the example of q2-feature-classifier.\n",
    "\n",
    "3) **[Classifier evaluation](./evaluate-classification.ipynb)** is performed based on taxonomic classifications generated by each classifier used in *step 2*. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [conda env:python-2.7]",
   "language": "python",
   "name": "conda-env-python-2.7-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
